packages/adblocker/src/preprocessor.ts

```diff
@@ -0,0 +1,241 @@
+  }
+
+  const command = line.slice(2);
+
+  if (fastStartsWith(command, 'if ')) {
+    return PreprocessorTypes.BEGIF;
+  }
+
+  if (command === 'else') {
+    return PreprocessorTypes.ELSE;
+  }
+
+  if (command === 'endif') {
+    return PreprocessorTypes.ENDIF;
+  }
+
+  return PreprocessorTypes.INVALID;
+}
+
+const tokenizerPattern = /(!|&&|\|\||\(|\)|[a-zA-Z0-9_]+)/g;
+
+const tokenize = (expression: string) => expression.match(tokenizerPattern);
+
+const precedence: Record<string, number> = {
+  '!': 2,
+  '&&': 1,
+  '||': 0,
+};
+
+const isOperator = (token: string) => Object.prototype.hasOwnProperty.call(precedence, token);
+
+/// The parsing is done using the [Shunting yard algorithm](https://en.wikipedia.org/wiki/Shunting_yard_algorithm).
+/// This function takes as input a string expression and an environment Map.
+/// The expression is made of constants (identifiers), logical operators
+/// (&&, ||), negations (!constant) and parentheses.
+///
+/// The environment is a simple Map that associates identifiers to boolean values.
+///
+/// The function should return the result of evaluating the expression using
+/// the values from `environment`. The return value of this function is
+/// either `true` or `false`.
+export const evaluate = (expression: string, env: Env) => {
+  const length = expression.length;
+
+  // Fast path for very simple conditions (which are at the moment 100% of all
+  // preprocessor conditions: either a single constant or a negated constant)
+  if (!length) {
+    return false;
+  }
+
+  const tokens = tokenize(expression);
+
+  if (!tokens) {
+    return false;
+  }
+
+  const output: (boolean | string)[] = [];
+  const stack: (boolean | string)[] = [];
+
+  let size = 0;
+
+  for (const token of tokens) {
+    size += token.length;
+
+    if (token === '(') {
+      stack.push(token);
+    } else if (token === ')') {
+      const parenthesesOpenIndex = stack.lastIndexOf('(');
+
+      // If the opening parenthesis doesn't exist
+      if (parenthesesOpenIndex < 0) {
+        return false;
+      }
+
+      output.push(...stack.splice(parenthesesOpenIndex + 1, stack.length - parenthesesOpenIndex));
+      stack.pop();
+    } else if (isOperator(token)) {
+      while (
+        stack.length &&
+        isOperator(stack[stack.length - 1] as string) &&
+        precedence[token] <= precedence[stack[stack.length - 1] as string]
+      ) {
+        output.push(stack.pop()!);
+      }
+
+      stack.push(token);
+    } else if (token === 'true') {
+      output.push(true);
+    } else if (token === 'false') {
+      output.push(false);
+    } else {
+      output.push(!!env.get(token));
+    }
+  }
+
+  // Exit if an unallowed character found.
+  // Since we're tokenizing via String.prototype.match function,
+  // the total length of matched tokens will be different in case
+  // unallowed characters were injected.
+  // However, we expect all spaces were already removed in prior step.
+  if (length !== size) {
```

Comment:
I think we can do this check before starting the evaluation (right after tokenizing).

```js
if (expression.length !== tokens.reduce((partialSum, token) => partialSum + token.length, 0)) {
  return false;
}
```

Reply:
Thanks. This has been applied to the current version.